src/source/agc.rs

```diff
@@ -0,0 +1,345 @@
+//
+//      Automatic Gain Control (AGC) Algorithm
+//      Designed by @UnknownSuperficialNight
+//
+//   Features:
+//   • Adaptive peak detection
+//   • RMS-based level estimation
+//   • Asymmetric attack/release
+//   • RMS-based general adjustments with peak limiting
+//
+//   Optimized for smooth and responsive gain control
+//
+//   Crafted with love. Enjoy! :)
+//
+
+use super::SeekError;
+use crate::{Sample, Source};
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::time::Duration;
+
+#[cfg(feature = "tracing")]
+use tracing;
+
+/// Size of the circular buffer used for RMS calculation.
+/// A larger size provides more stable RMS values but increases latency.
+const RMS_WINDOW_SIZE: usize = 8192;
+
+/// Automatic Gain Control filter for maintaining consistent output levels.
+///
+/// This struct implements an AGC algorithm that dynamically adjusts audio levels
+/// based on both peak and RMS (Root Mean Square) measurements.
+#[derive(Clone, Debug)]
+pub struct AutomaticGainControl<I> {
+    input: I,
+    target_level: f32,
+    absolute_max_gain: f32,
+    current_gain: f32,
+    attack_coeff: f32,
+    release_coeff: f32,
+    min_attack_coeff: f32,
+    peak_level: f32,
+    rms_window: CircularBuffer,
+    is_enabled: Arc<AtomicBool>,
+}
+
+/// A circular buffer for efficient RMS calculation over a sliding window.
+///
+/// This structure allows for constant-time updates and mean calculations,
+/// which is crucial for real-time audio processing.
+#[derive(Clone, Debug)]
+struct CircularBuffer {
+    buffer: [f32; RMS_WINDOW_SIZE],
```

Comment:
32Kb is a rather large allocation on the stack. I am kinda surprised that clippy is not complaining here. Might be better to heap allocate this (vec<f32>). You could also try replacing buffer+index in `CircularBuffer` with [`VecDeque`](https://doc.rust-lang.org/std/collections/struct.VecDeque.html#). But that might have some overhead. Since you added a benchmark we can measure that now :stuck_out_tongue_closed_eyes:.

Reply:
Tried VecDeque added 10ms, upped mean by 4ms

![image](https://github.com/user-attachments/assets/94f9c519-67c8-47a5-9dd1-d0e9eedaf5bf)

Reply:
and putting the buffer on the heap with Box or Vec? Does that do the same?

Reply:
Fixed in this [commit](https://github.com/RustAudio/rodio/pull/621/commits/86cb156e47bb0165a8463e75c70b284903189337)

Reply:
> and putting the buffer on the heap with Box or Vec? Does that do the same? Did you pre-allocate the VecDeque during source creation? (filling it up with zero's)?

Sorry didn't see the reply before the commit yea I did not pre-allocate the VecDeque

here is pre alloc VecDeque
├─ agc_enabled   13.83 ms      │ 15.11 ms      │ 13.91 ms      │ 13.95 ms      │ 100     │ 100

here is using the commit I made with box
├─ agc_enabled   8.682 ms      │ 12.19 ms      │ 9.479 ms      │ 9.564 ms      │ 100     │ 100

still a quite bit worse